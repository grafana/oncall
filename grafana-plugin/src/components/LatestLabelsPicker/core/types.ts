export interface ItemRepresentation {
  id: string;
  repr: string;
}

export interface ItemGroup {
  key: ItemRepresentation;
  values: ItemRepresentation[];
}

export interface ItemSelected {
  key: ItemRepresentation;
  value: ItemRepresentation;
}