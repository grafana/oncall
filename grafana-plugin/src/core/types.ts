export interface ItemGroup {
  key: any;
  values: any[];
}

export interface ItemSelected {
  key: any;
  value: any;
}

export enum LabelInputType {
  key = 'key',
  value = 'value',
}

export interface ServiceLabelValidator {
  isValid: boolean;
  errorMessage?: string;
}
