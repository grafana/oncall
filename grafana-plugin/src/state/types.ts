import { AppPluginMeta, KeyValue } from '@grafana/data';

import { RootStore } from 'state/rootStore';

export interface WithStoreProps {
  store: RootStore;
}

export interface PageProps<T extends KeyValue = KeyValue> {
  meta: AppPluginMeta<T>;
  query: KeyValue;
}

export interface SelectOption {
  value: string | number;
  display_name: string;
}
