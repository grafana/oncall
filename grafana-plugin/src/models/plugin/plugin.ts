import { isEqual } from 'lodash-es';
import { makeAutoObservable, runInAction } from 'mobx';
import { OnCallPluginMetaJSONData } from 'types';

import { ActionKey } from 'models/loader/action-keys';
import { GrafanaApiClient } from 'network/grafana-api/http-client';
import { makeRequest } from 'network/network';
import { PluginConnection, StatusResponse } from 'network/oncall-api/api.types';
import { RootBaseStore } from 'state/rootBaseStore/RootBaseStore';
import { waitInMs } from 'utils/async';
import { AutoLoadingState } from 'utils/decorators';

import { PluginHelper } from './plugin.helper';

/* 
High-level OnCall initialization process:
On OSS:
  - On OnCall page / OnCall extension mount POST /status is called and it has pluginConfiguration object with different flags. 
    If all of them have `ok: true` , we consider plugin to be successfully configured and application loading is being continued. 
    Otherwise, we show error page with the option to go to plugin config (for Admin user) or to contact administrator (for nonAdmin user)
  - On plugin config page frontend sends another POST /status. If every flag has `ok: true`, it shows that plugin is connected. 
    Otherwise, it shows more detailed information of what is misconfigured / missing. User can update onCallApiUrl and try to reconnect plugin.
      - If Grafana version >= 10.3 AND externalServiceAccount feature flag is `true`, then grafana token is autoprovisioned and there is no need to create it
      - Otherwise, user is given the option to manually create service account as Admin and then reconnect the plugin
On Cloud:
  - On OnCall page / OnCall extension mount POST /status is called. If plugin is configured correctly, application loads as usual.
    If it's not, we show error page with the button to contact support
  - On plugin config page we show info if plugin is connected. If it's not we show detailed information of the errors and the button to contact support
*/

export class PluginStore {
  rootStore: RootBaseStore;
  connectionStatus?: PluginConnection;
  apiUrlFromStatus?: string;
  isPluginConnected = false;
  appliedOnCallApiUrl = '';

  constructor(rootStore: RootBaseStore) {
    makeAutoObservable(this, undefined, { autoBind: true });
    this.rootStore = rootStore;
  }

  private resetConnectionStatus() {
    this.connectionStatus = undefined;
    this.isPluginConnected = false;
  }

  async refreshAppliedOnCallApiUrl() {
    const { jsonData } = await GrafanaApiClient.getGrafanaPluginSettings();
    runInAction(() => {
      this.appliedOnCallApiUrl = jsonData.onCallApiUrl;
    });
  }

  @AutoLoadingState(ActionKey.PLUGIN_VERIFY_CONNECTION)
  async verifyPluginConnection() {
    const { pluginConnection, api_url } = await makeRequest<StatusResponse>(`/plugin/status`, {});
    runInAction(() => {
      this.connectionStatus = pluginConnection;
      this.apiUrlFromStatus = api_url;
      this.isPluginConnected = Object.keys(pluginConnection).every(
        (key) => pluginConnection[key as keyof PluginConnection]?.ok
      );
    });
  }

  @AutoLoadingState(ActionKey.PLUGIN_UPDATE_SETTINGS_AND_REINITIALIZE)
  async updatePluginSettingsAndReinitializePlugin({
    currentJsonData,
    newJsonData,
  }: {
    currentJsonData: OnCallPluginMetaJSONData;
    newJsonData: Partial<OnCallPluginMetaJSONData>;
  }) {
    this.resetConnectionStatus();
    const saveJsonDataCandidate = { ...currentJsonData, ...newJsonData };
    if (!isEqual(currentJsonData, saveJsonDataCandidate) || !this.connectionStatus?.oncall_api_url?.ok) {
      await GrafanaApiClient.updateGrafanaPluginSettings({ jsonData: saveJsonDataCandidate });
      await waitInMs(1000); // It's required for backend proxy to pick up new settings
    }
    try {
      await PluginHelper.install();
    } finally {
      await this.verifyPluginConnection();
    }
  }

  @AutoLoadingState(ActionKey.PLUGIN_RECREATE_SERVICE_ACCOUNT)
  async recreateServiceAccountAndRecheckPluginStatus() {
    await GrafanaApiClient.recreateGrafanaTokenAndSaveInPluginSettings();
    await this.verifyPluginConnection();
  }

  async enablePlugin() {
    await GrafanaApiClient.updateGrafanaPluginSettings({}, true);
    location.reload();
  }
}
