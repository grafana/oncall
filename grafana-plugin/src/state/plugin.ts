import { getBackendSrv } from '@grafana/runtime';

import { makeRequest } from 'network';

export async function createGrafanaToken() {
  const keys = await getBackendSrv().get('/api/auth/keys');
  const existingKey = keys.find((key: { id: number; name: string; role: string }) => key.name === 'OnCall');

  if (existingKey) {
    await getBackendSrv().delete(`/api/auth/keys/${existingKey.id}`);
  }

  return await getBackendSrv().post('/api/auth/keys', {
    name: 'OnCall',
    role: 'Admin',
    secondsToLive: null,
  });
}

export async function updateGrafanaToken(key: string) {
  await getBackendSrv().post(`/api/plugins/grafana-oncall-app/settings`, {
    enabled: true,
    pinned: true,
    secureJsonData: {
      grafanaToken: key,
    },
  });
}

export async function startPluginSync() {
  return await makeRequest('/plugin/sync', { method: 'POST' });
}

export const SYNC_STATUS_RETRY_LIMIT = 10;

export const syncStatusDelay = (retryCount) => new Promise((resolve) => setTimeout(resolve, 10 * 2 ** retryCount));

export async function getPluginSyncStatus() {
  return await makeRequest(`/plugin/sync`, { method: 'GET' });
}

export async function installPlugin() {
  const grafanaToken = await createGrafanaToken();
  await updateGrafanaToken(grafanaToken.key);
  return await makeRequest('/plugin/install', { method: 'POST' });
}
