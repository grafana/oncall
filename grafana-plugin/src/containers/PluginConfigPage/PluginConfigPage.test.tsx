import React from 'react';

import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { useLocation as useLocationOriginal } from 'react-router-dom';
import { OnCallPluginConfigPageProps } from 'types';

import { PluginState } from 'state/plugin/plugin';

import {
  PluginConfigPage,
  reloadPageWithPluginConfiguredQueryParams,
  removePluginConfiguredQueryParams,
} from './PluginConfigPage';

jest.mock('../../../package.json', () => ({
  version: 'v1.2.3',
}));

jest.mock('react-router-dom', () => ({
  useLocation: jest.fn(() => ({
    search: '',
  })),
}));

const useLocation = useLocationOriginal as jest.Mock<ReturnType<typeof useLocationOriginal>>;

enum License {
  OSS = 'OpenSource',
  CLOUD = 'some-other-license',
}

const CHECK_IF_PLUGIN_IS_CONNECTED_ERROR_MESSAGE = 'ohhh nooo a plugin connection error';
const UPDATE_PLUGIN_STATUS_ERROR_MESSAGE = 'ohhh noooo a sync issue';
const PLUGIN_CONFIGURATION_FORM_DATA_ID = 'plugin-configuration-form';
const STATUS_MESSAGE_BLOCK_DATA_ID = 'status-message-block';

const MOCK_PROTOCOL = 'https:';
const MOCK_HOST = 'localhost:3000';
const MOCK_PATHNAME = '/dkjdfjkfd';
const MOCK_URL = `${MOCK_PROTOCOL}//${MOCK_HOST}${MOCK_PATHNAME}`;

/**
 * this is just a little hack to silence a warning that we'll get until we
 * upgrade to 16.9. See also: https://github.com/facebook/react/pull/14853
 * https://github.com/testing-library/react-testing-library#suppressing-unnecessary-warnings-on-react-dom-168
 */
const originalError = console.error;

beforeEach(() => {
  delete global.window.location;
  global.window ??= Object.create(window);
  global.window.location = {
    protocol: MOCK_PROTOCOL,
    host: MOCK_HOST,
    pathname: MOCK_PATHNAME,
    href: MOCK_URL,
  } as Location;
  global.window.history.pushState = jest.fn();

  console.error = (...args) => {
    if (/Warning.*not wrapped in act/.test(args[0])) {
      return;
    }
    originalError.call(console, ...args);
  };
});

afterEach(() => {
  jest.clearAllMocks();
  console.error = originalError;
});

const mockCheckTokenAndIfPluginIsConnected = (license: License = License.OSS) => {
  PluginState.checkTokenAndIfPluginIsConnected = jest.fn().mockResolvedValueOnce({
    token_ok: true,
    license,
    version: 'v1.2.3',
    allow_signup: true,
    currently_undergoing_maintenance_message: null,
    recaptcha_site_key: 'abc',
    is_installed: true,
    is_user_anonymous: false,
  });
};

const generateComponentProps = (
  onCallApiUrl: OnCallPluginConfigPageProps['plugin']['meta']['jsonData']['onCallApiUrl'] = null,
  enabled = false
): OnCallPluginConfigPageProps =>
  ({
    plugin: {
      meta: {
        jsonData: onCallApiUrl === null ? null : { onCallApiUrl },
        enabled,
      },
    },
  } as OnCallPluginConfigPageProps);

describe('reloadPageWithPluginConfiguredQueryParams', () => {
  test.each([true, false])(
    'it modifies the query params depending on whether or not the plugin is already enabled: enabled - %s',
    (pluginEnabled) => {
      // mocks
      const version = 'v1.2.3';
      const license = 'OpenSource';
      const recaptcha_site_key = 'abc';
      const currently_undergoing_maintenance_message = 'false';

      // test
      reloadPageWithPluginConfiguredQueryParams(
        { version, license, recaptcha_site_key, currently_undergoing_maintenance_message },
        pluginEnabled
      );

      // assertions
      expect(window.location.href).toEqual(
        pluginEnabled
          ? MOCK_URL
          : `${MOCK_URL}?pluginConfigured=true&pluginConfiguredLicense=${license}&pluginConfiguredVersion=${version}`
      );
    }
  );
});

describe('removePluginConfiguredQueryParams', () => {
  test('it removes all the query params if history.pushState is available, and plugin is enabled', () => {
    removePluginConfiguredQueryParams(true);
    expect(window.history.pushState).toHaveBeenCalledWith({ path: MOCK_URL }, '', MOCK_URL);
  });

  test('it does not remove all the query params if history.pushState is available, and plugin is disabled', () => {
    removePluginConfiguredQueryParams(false);
    expect(window.history.pushState).not.toHaveBeenCalled();
  });
});

describe('PluginConfigPage', () => {
  test('It removes the plugin configured query params if the plugin is enabled', async () => {
    // mocks
    const metaJsonDataOnCallApiUrl = 'onCallApiUrlFromMetaJsonData';
    PluginState.updatePluginStatus = jest.fn();
    mockCheckTokenAndIfPluginIsConnected();

    // test setup
    render(<PluginConfigPage {...generateComponentProps(metaJsonDataOnCallApiUrl, true)} />);
    await screen.findByTestId(STATUS_MESSAGE_BLOCK_DATA_ID);

    // assertions
    expect(window.history.pushState).toHaveBeenCalledWith({ path: MOCK_URL }, '', MOCK_URL);

    expect(PluginState.updatePluginStatus).toHaveBeenCalledTimes(1);
    expect(PluginState.updatePluginStatus).toHaveBeenCalledWith(metaJsonDataOnCallApiUrl);

    expect(PluginState.checkTokenAndIfPluginIsConnected).toHaveBeenCalledTimes(1);
    expect(PluginState.checkTokenAndIfPluginIsConnected).toHaveBeenCalledWith(metaJsonDataOnCallApiUrl);
  });

  test("It doesn't make any network calls if the plugin configured query params are provided", async () => {
    // mocks
    const metaJsonDataOnCallApiUrl = 'onCallApiUrlFromMetaJsonData';
    const version = 'v1.2.3';
    const license = 'OpenSource';

    useLocation.mockReturnValueOnce({
      search: `?pluginConfigured=true&pluginConfiguredLicense=${license}&pluginConfiguredVersion=${version}`,
    } as ReturnType<typeof useLocationOriginal>);

    PluginState.updatePluginStatus = jest.fn();
    mockCheckTokenAndIfPluginIsConnected();

    // test setup
    const component = render(<PluginConfigPage {...generateComponentProps(metaJsonDataOnCallApiUrl)} />);
    await screen.findByTestId(STATUS_MESSAGE_BLOCK_DATA_ID);

    // assertions
    expect(PluginState.updatePluginStatus).not.toHaveBeenCalled();
    expect(PluginState.checkTokenAndIfPluginIsConnected).not.toHaveBeenCalled();
    expect(component.container).toMatchSnapshot();
  });

  test("If onCallApiUrl is not set in the plugin's meta jsonData, or in process.env, updatePluginStatus is not called, and the configuration form is shown", async () => {
    // mocks
    delete process.env.ONCALL_API_URL;

    PluginState.updatePluginStatus = jest.fn();
    PluginState.checkTokenAndIfPluginIsConnected = jest.fn();

    // test setup
    const component = render(<PluginConfigPage {...generateComponentProps()} />);
    await screen.findByTestId(PLUGIN_CONFIGURATION_FORM_DATA_ID);

    // assertions
    expect(PluginState.updatePluginStatus).not.toHaveBeenCalled();
    expect(PluginState.checkTokenAndIfPluginIsConnected).not.toHaveBeenCalled();
    expect(component.container).toMatchSnapshot();
  });

  test('If onCallApiUrl is set, and updatePluginStatus returns an error, it sets an error message', async () => {
    // mocks
    const processEnvOnCallApiUrl = 'onCallApiUrlFromProcessEnv';
    const metaJsonDataOnCallApiUrl = 'onCallApiUrlFromMetaJsonData';

    process.env.ONCALL_API_URL = processEnvOnCallApiUrl;

    PluginState.updatePluginStatus = jest.fn().mockResolvedValueOnce(CHECK_IF_PLUGIN_IS_CONNECTED_ERROR_MESSAGE);

    // test setup
    const component = render(<PluginConfigPage {...generateComponentProps(metaJsonDataOnCallApiUrl)} />);
    await screen.findByTestId(STATUS_MESSAGE_BLOCK_DATA_ID);

    // assertions
    expect(PluginState.updatePluginStatus).toHaveBeenCalledTimes(1);
    expect(PluginState.updatePluginStatus).toHaveBeenCalledWith(metaJsonDataOnCallApiUrl);
    expect(component.container).toMatchSnapshot();
  });

  test('OnCallApiUrl is set, and checkTokenAndIfPluginIsConnected returns an error', async () => {
    // mocks
    const processEnvOnCallApiUrl = 'onCallApiUrlFromProcessEnv';
    const metaJsonDataOnCallApiUrl = 'onCallApiUrlFromMetaJsonData';

    process.env.ONCALL_API_URL = processEnvOnCallApiUrl;

    PluginState.updatePluginStatus = jest.fn().mockResolvedValueOnce(null);
    PluginState.checkTokenAndIfPluginIsConnected = jest.fn().mockResolvedValueOnce(UPDATE_PLUGIN_STATUS_ERROR_MESSAGE);

    // test setup
    const component = render(<PluginConfigPage {...generateComponentProps(metaJsonDataOnCallApiUrl)} />);
    await screen.findByTestId(STATUS_MESSAGE_BLOCK_DATA_ID);

    // assertions
    expect(PluginState.updatePluginStatus).toHaveBeenCalledTimes(1);
    expect(PluginState.updatePluginStatus).toHaveBeenCalledWith(metaJsonDataOnCallApiUrl);
    expect(component.container).toMatchSnapshot();
  });

  test.each([License.CLOUD, License.OSS])(
    'OnCallApiUrl is set, and checkTokenAndIfPluginIsConnected does not return an error. It displays properly the plugin connected items based on the license - License: %s',
    async (license) => {
      // mocks
      const processEnvOnCallApiUrl = 'onCallApiUrlFromProcessEnv';
      const metaJsonDataOnCallApiUrl = 'onCallApiUrlFromMetaJsonData';

      process.env.ONCALL_API_URL = processEnvOnCallApiUrl;

      PluginState.updatePluginStatus = jest.fn().mockResolvedValueOnce(null);
      mockCheckTokenAndIfPluginIsConnected(license);

      // test setup
      const component = render(<PluginConfigPage {...generateComponentProps(metaJsonDataOnCallApiUrl)} />);
      await screen.findByTestId(STATUS_MESSAGE_BLOCK_DATA_ID);

      // assertions
      expect(PluginState.updatePluginStatus).toHaveBeenCalledTimes(1);
      expect(PluginState.updatePluginStatus).toHaveBeenCalledWith(metaJsonDataOnCallApiUrl);
      expect(component.container).toMatchSnapshot();
    }
  );

  test.each([true, false])('Plugin reset: successful - %s', async (successful) => {
    // mocks
    const processEnvOnCallApiUrl = 'onCallApiUrlFromProcessEnv';
    const metaJsonDataOnCallApiUrl = 'onCallApiUrlFromMetaJsonData';

    process.env.ONCALL_API_URL = processEnvOnCallApiUrl;
    window.location.reload = jest.fn();

    PluginState.updatePluginStatus = jest.fn().mockResolvedValue(null);
    mockCheckTokenAndIfPluginIsConnected(License.OSS);

    if (successful) {
      PluginState.resetPlugin = jest.fn().mockResolvedValueOnce(null);
    } else {
      PluginState.resetPlugin = jest.fn().mockRejectedValueOnce('dfdf');
    }

    // test setup
    const component = render(<PluginConfigPage {...generateComponentProps(metaJsonDataOnCallApiUrl)} />);
    const button = await screen.findByRole('button');

    // click the reset button, which opens the modal
    await userEvent.click(button);
    // click the confirm button within the modal, which actually triggers the callback
    await userEvent.click(screen.getByText('Remove'));

    // assertions
    expect(PluginState.updatePluginStatus).toHaveBeenCalledTimes(1);
    expect(PluginState.updatePluginStatus).toHaveBeenCalledWith(metaJsonDataOnCallApiUrl);

    expect(PluginState.checkTokenAndIfPluginIsConnected).toHaveBeenCalledTimes(1);
    expect(PluginState.checkTokenAndIfPluginIsConnected).toHaveBeenCalledWith(metaJsonDataOnCallApiUrl);

    expect(PluginState.resetPlugin).toHaveBeenCalledTimes(1);
    expect(PluginState.resetPlugin).toHaveBeenCalledWith();

    expect(component.container).toMatchSnapshot();
  });
});
