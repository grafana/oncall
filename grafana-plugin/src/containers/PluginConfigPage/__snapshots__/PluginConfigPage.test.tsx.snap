// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PluginConfigPage If onCallApiUrl is not set in the plugin's meta jsonData, and ONCALL_API_URL is passed in process.env, and there is an error calling selfHostedInstallPlugin, it sets an error message 1`] = `
<div>
  <legend
    class="css-1w9pvsj"
  >
    Configure Grafana OnCall
  </legend>
  <p>
    This page will help you configure the OnCall plugin ðŸ‘‹
  </p>
  <pre
    data-testid="status-message-block"
  >
    <span
      class="root text text--undefined text--medium"
    >
      ohhh nooo an error msg from self hosted install plugin
    </span>
  </pre>
  <div
    class="css-ffyaiw-horizontal-group"
    style="width: 100%; height: 100%;"
  >
    <div
      class="css-18qv8yz-layoutChildrenWrapper"
    >
      <button
        class="css-td06pi-button"
        type="button"
      >
        <span
          class="css-1riaxdn"
        >
          Retry Sync
        </span>
      </button>
    </div>
    <div
      class="css-18qv8yz-layoutChildrenWrapper"
    >
      <button
        class="css-ttl745-button"
        type="button"
      >
        <span
          class="css-1riaxdn"
        >
          Remove current configuration
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`PluginConfigPage If onCallApiUrl is not set in the plugin's meta jsonData, or in process.env, updatePluginStatus is not called, and the configuration form is shown 1`] = `
<div>
  <legend
    class="css-1w9pvsj"
  >
    Configure Grafana OnCall
  </legend>
  <p>
    This page will help you configure the OnCall plugin ðŸ‘‹
  </p>
  <form
    class="css-1srg48i"
    data-testid="plugin-configuration-form"
  >
    <div
      class="info-block"
    >
      <p>
        1. Launch the OnCall backend
      </p>
      <span
        class="root text text--secondary text--medium"
      >
        Run hobby, dev or production backend. See
         
        <a
          href="https://github.com/grafana/oncall#getting-started"
          rel="noreferrer"
          target="_blank"
        >
          <span
            class="root text text--link text--medium"
          >
            here
          </span>
        </a>
         
        on how to get started.
      </span>
    </div>
    <div
      class="info-block"
    >
      <p>
        2. Let us know the base URL of your OnCall API
      </p>
      <span
        class="root text text--secondary text--medium"
      >
        The OnCall backend must be reachable from your Grafana installation. Some examples are:
        <br />
        - http://host.docker.internal:8080
        <br />
        - http://localhost:8080
      </span>
    </div>
    <div
      class="css-1rplq84"
    >
      <div
        class="css-l4ykjo-Label"
      >
        <label>
          <div
            class="css-70qvj9"
          >
            OnCall backend URL
          </div>
        </label>
      </div>
      <div>
        <div>
          <div
            class="css-1cvb0sk-input-wrapper"
            data-testid="input-wrapper"
          >
            <div
              class="css-10lnb82-input-inputWrapper"
            >
              <input
                class="css-8tk2dk-input-input"
                data-testid="onCallApiUrl"
                name="onCallApiUrl"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <button
      class="css-td06pi-button"
      type="submit"
    >
      <span
        class="css-1riaxdn"
      >
        Connect
      </span>
    </button>
  </form>
</div>
`;

exports[`PluginConfigPage If onCallApiUrl is set, and updatePluginStatus returns an error, it sets an error message 1`] = `
<div>
  <legend
    class="css-1w9pvsj"
  >
    Configure Grafana OnCall
  </legend>
  <p>
    This page will help you configure the OnCall plugin ðŸ‘‹
  </p>
  <pre
    data-testid="status-message-block"
  >
    <span
      class="root text text--undefined text--medium"
    >
      ohhh nooo a plugin connection error
    </span>
  </pre>
  <div
    class="css-ffyaiw-horizontal-group"
    style="width: 100%; height: 100%;"
  >
    <div
      class="css-18qv8yz-layoutChildrenWrapper"
    >
      <button
        class="css-td06pi-button"
        type="button"
      >
        <span
          class="css-1riaxdn"
        >
          Retry Sync
        </span>
      </button>
    </div>
    <div
      class="css-18qv8yz-layoutChildrenWrapper"
    >
      <button
        class="css-ttl745-button"
        type="button"
      >
        <span
          class="css-1riaxdn"
        >
          Remove current configuration
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`PluginConfigPage It doesn't make any network calls if the plugin configured query params are provided 1`] = `
<div>
  <legend
    class="css-1w9pvsj"
  >
    Configure Grafana OnCall
  </legend>
  <pre
    data-testid="status-message-block"
  >
    <span
      class="root text text--undefined text--medium"
    >
      Connected to OnCall (v1.2.3, OpenSource)
    </span>
  </pre>
  <div
    class="css-ffyaiw-horizontal-group"
    style="width: 100%; height: 100%;"
  >
    <div
      class="css-18qv8yz-layoutChildrenWrapper"
    >
      <a
        class="css-td06pi-button"
        href="/a/grafana-oncall-app/"
        tabindex="0"
      >
        <span
          class="css-1riaxdn"
        >
          Open Grafana OnCall
        </span>
      </a>
    </div>
    <div
      class="css-18qv8yz-layoutChildrenWrapper"
    >
      <button
        class="css-ttl745-button"
        type="button"
      >
        <span
          class="css-1riaxdn"
        >
          Remove current configuration
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`PluginConfigPage OnCallApiUrl is set, and checkTokenAndIfPluginIsConnected does not return an error. It displays properly the plugin connected items based on the license - License: OpenSource 1`] = `
<div>
  <legend
    class="css-1w9pvsj"
  >
    Configure Grafana OnCall
  </legend>
  <pre
    data-testid="status-message-block"
  >
    <span
      class="root text text--undefined text--medium"
    >
      Connected to OnCall (v1.2.3, OpenSource)
    </span>
  </pre>
  <div
    class="css-ffyaiw-horizontal-group"
    style="width: 100%; height: 100%;"
  >
    <div
      class="css-18qv8yz-layoutChildrenWrapper"
    >
      <a
        class="css-td06pi-button"
        href="/a/grafana-oncall-app/"
        tabindex="0"
      >
        <span
          class="css-1riaxdn"
        >
          Open Grafana OnCall
        </span>
      </a>
    </div>
    <div
      class="css-18qv8yz-layoutChildrenWrapper"
    >
      <button
        class="css-ttl745-button"
        type="button"
      >
        <span
          class="css-1riaxdn"
        >
          Remove current configuration
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`PluginConfigPage OnCallApiUrl is set, and checkTokenAndIfPluginIsConnected does not return an error. It displays properly the plugin connected items based on the license - License: some-other-license 1`] = `
<div>
  <legend
    class="css-1w9pvsj"
  >
    Configure Grafana OnCall
  </legend>
  <pre
    data-testid="status-message-block"
  >
    <span
      class="root text text--undefined text--medium"
    >
      Connected to OnCall (v1.2.3, some-other-license)
    </span>
  </pre>
  <div
    class="css-gjl87o-vertical-group"
    style="width: 100%; height: 100%;"
  >
    <div
      class="css-gxt817-layoutChildrenWrapper"
    >
      <div
        class="css-l4ykjo-Label"
      >
        <label>
          <div
            class="css-70qvj9"
          >
            This is a cloud managed configuration.
          </div>
        </label>
      </div>
    </div>
    <div
      class="css-gxt817-layoutChildrenWrapper"
    >
      <a
        class="css-td06pi-button"
        href="/a/grafana-oncall-app/"
        tabindex="0"
      >
        <span
          class="css-1riaxdn"
        >
          Open Grafana OnCall
        </span>
      </a>
    </div>
  </div>
</div>
`;

exports[`PluginConfigPage OnCallApiUrl is set, and checkTokenAndIfPluginIsConnected returns an error 1`] = `
<div>
  <legend
    class="css-1w9pvsj"
  >
    Configure Grafana OnCall
  </legend>
  <p>
    This page will help you configure the OnCall plugin ðŸ‘‹
  </p>
  <pre
    data-testid="status-message-block"
  >
    <span
      class="root text text--undefined text--medium"
    >
      ohhh noooo a sync issue
    </span>
  </pre>
  <div
    class="css-ffyaiw-horizontal-group"
    style="width: 100%; height: 100%;"
  >
    <div
      class="css-18qv8yz-layoutChildrenWrapper"
    >
      <button
        class="css-td06pi-button"
        type="button"
      >
        <span
          class="css-1riaxdn"
        >
          Retry Sync
        </span>
      </button>
    </div>
    <div
      class="css-18qv8yz-layoutChildrenWrapper"
    >
      <button
        class="css-ttl745-button"
        type="button"
      >
        <span
          class="css-1riaxdn"
        >
          Remove current configuration
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`PluginConfigPage Plugin reset: successful - false 1`] = `
<div>
  <legend
    class="css-1w9pvsj"
  >
    Configure Grafana OnCall
  </legend>
  <p>
    This page will help you configure the OnCall plugin ðŸ‘‹
  </p>
  <pre
    data-testid="status-message-block"
  >
    <span
      class="root text text--undefined text--medium"
    >
      There was an error resetting your plugin, try again.
    </span>
  </pre>
  <div
    class="css-ffyaiw-horizontal-group"
    style="width: 100%; height: 100%;"
  >
    <div
      class="css-18qv8yz-layoutChildrenWrapper"
    >
      <button
        class="css-td06pi-button"
        type="button"
      >
        <span
          class="css-1riaxdn"
        >
          Retry Sync
        </span>
      </button>
    </div>
    <div
      class="css-18qv8yz-layoutChildrenWrapper"
    >
      <button
        class="css-ttl745-button"
        type="button"
      >
        <span
          class="css-1riaxdn"
        >
          Remove current configuration
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`PluginConfigPage Plugin reset: successful - true 1`] = `
<div>
  <legend
    class="css-1w9pvsj"
  >
    Configure Grafana OnCall
  </legend>
  <p>
    This page will help you configure the OnCall plugin ðŸ‘‹
  </p>
  <form
    class="css-1srg48i"
    data-testid="plugin-configuration-form"
  >
    <div
      class="info-block"
    >
      <p>
        1. Launch the OnCall backend
      </p>
      <span
        class="root text text--secondary text--medium"
      >
        Run hobby, dev or production backend. See
         
        <a
          href="https://github.com/grafana/oncall#getting-started"
          rel="noreferrer"
          target="_blank"
        >
          <span
            class="root text text--link text--medium"
          >
            here
          </span>
        </a>
         
        on how to get started.
      </span>
    </div>
    <div
      class="info-block"
    >
      <p>
        2. Let us know the base URL of your OnCall API
      </p>
      <span
        class="root text text--secondary text--medium"
      >
        The OnCall backend must be reachable from your Grafana installation. Some examples are:
        <br />
        - http://host.docker.internal:8080
        <br />
        - http://localhost:8080
      </span>
    </div>
    <div
      class="css-1rplq84"
    >
      <div
        class="css-l4ykjo-Label"
      >
        <label>
          <div
            class="css-70qvj9"
          >
            OnCall backend URL
          </div>
        </label>
      </div>
      <div>
        <div>
          <div
            class="css-1cvb0sk-input-wrapper"
            data-testid="input-wrapper"
          >
            <div
              class="css-10lnb82-input-inputWrapper"
            >
              <input
                class="css-8tk2dk-input-input"
                data-testid="onCallApiUrl"
                name="onCallApiUrl"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <button
      class="css-td06pi-button"
      type="submit"
    >
      <span
        class="css-1riaxdn"
      >
        Connect
      </span>
    </button>
  </form>
</div>
`;
